### PR TITLE
[FLINK-25371] Include data port as part of the host info for subtask detail panel on Web UI

### PR DESCRIPTION
**Pull request by X-czh**
## What is the purpose of the change

Help user locate the container where a task is running on from Web UI by including port in the host info on Web UI.

## Brief change log

1. Include port as part of the host info in `SubtaskExecutionAttemptDetailsInfo`.
2. Introduce a new method `toLocationString` in `TaskManagerLocation` to align the behavior of pretty printing taskmanager location in REST handlers.
3. Rename column name from "Host" to "Host:port" on Web UI to reflect the change of adding port info.

## Verifying this change

This change is already covered by existing tests, such as `SubtaskCurrentAttemptDetailsHandlerTest.testHandleRequest()`.

Also manually verified the change by running a cluster.

image

image

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) yes
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) not documented

### CODE REVIEW

**Comment by X-czh** [comment]
@flinkbot run azure

**Comment by huwh** [comment]
The screenshot displays the task manager panel. You may need to upload the subtasks information.

**Comment by X-czh** [comment]
> The screenshot displays the task manager panel. You may need to upload the subtasks information.

Thanks for the comment. I've updated the MR description with the screenshot of the subtask info page.

**Comment by X-czh** [comment]
@huwh Could you help review it again when you are free? Much appreciated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java [hunk 1]
```diff
@@ -203,7 +203,10 @@ public static SubtaskExecutionAttemptDetailsInfo create(
         final long now = System.currentTimeMillis();
 
         final TaskManagerLocation location = execution.getAssignedResourceLocation();
-        final String locationString = location == null ? "(unassigned)" : location.getHostname();
+        final String locationString =
```
Comment (huwh) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:206]: I've found that the API handles location in different ways, such as JobExceptionsHandler,JobVertexTaskManagersHandler, they connect the hostname and data port. 
But for SubtasksAllAccumulatorsHandler,SubtasksTimesHandler,SubtaskExecutionAttemptDetailsInfo, they only use the hostname. 

I think it's better to align these behaviors. We can extract the logic of the concat of hostname and port to a method of TaskManagerLocation.

**Review by huwh — COMMENTED** [review]
Thanks for the PR. I left some comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java (hunk 1 above)
Comment (X-czh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:206]: Thanks for the comment, I'll align the location generation logic in different places.

**Review by 1996fanrui — COMMENTED** [review]
Hi @X-czh , thanks for your contribution.

From your screenshot, we can see the title is host instead of `host:port`. However, the port is added.

So should we add the port to the title name as well?

**Comment by X-czh** [comment]
> location

Hi @1996fanrui, Thanks for the suggestion and I agree with that. How about using "Host:port" in the title in both the subtask details panel and tasks manager details panel?

**Comment by 1996fanrui** [comment]
> How about using "Host:port" in the title in both the subtask details panel and tasks manager details panel?

Sounds make sense to me.

**Comment by X-czh** [comment]
@huwh @1996fanrui I've updated the PR with two new commits that aligns the behavior of pretty printing taskmanager locations and updates the Web UI accordingly. Please help review again when you are free, thanks~

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/SubtaskCurrentAttemptDetailsHandlerTest.java [hunk 2]
```diff
@@ -186,7 +186,9 @@ public void testHandleRequest() throws Exception {
                         subtaskIndex,
                         expectedState,
                         attempt,
-                        assignedResourceLocation.getHostname(),
+                        assignedResourceLocation.getHostname()
```
Comment (huwh) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/SubtaskCurrentAttemptDetailsHandlerTest.java:189]: assignedResourceLocation.getLocationString()

**Review by huwh — COMMENTED** [review]
Thanks @X-czh, LGTM with a comment.

**Review by 1996fanrui — COMMENTED** [review]
Hi @X-czh thanks for the update.

I checked this PR locally, and found the `JobExceptionsHandler#toString(@Nullable TaskManagerLocation location)` should use the `getLocationString` instead of `location.getFQDNHostname(), location.dataPort()` as well.

Also, the `ArchivedTaskManagerLocation` is similar to `TaskManagerLocation`, and caller needs to combine the `fqdnHostname` and `dataPort` as well. Should we adding the `getLocationString()` to the `ArchivedTaskManagerLocation` as well? Or just keeping `locationString` for  `ArchivedTaskManagerLocation` is enough?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexTaskManagersHandler.java [hunk 3]
```diff
@@ -142,9 +142,7 @@ private static JobVertexTaskManagersInfo createJobVertexTaskManagersInfo(
             for (AccessExecution execution : vertex.getCurrentExecutions()) {
                 TaskManagerLocation location = execution.getAssignedResourceLocation();
                 String taskManagerHost =
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexTaskManagersHandler.java:144]:
```suggestion
                String locationString =
```

how about this?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java [hunk 4]
```diff
@@ -275,6 +275,15 @@ public static String getHostName(InetAddress inetAddress) {
         return hostName;
     }
 
+    /**
+     * Gets the location string of the TaskManager in the format of "$HOST:$PORT".
+     *
+     * @return The location string of the TaskManager.
+     */
+    public String getLocationString() {
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java:283]: Could you help add some tests for this method?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java [hunk 5]
```diff
@@ -203,7 +203,8 @@ public static SubtaskExecutionAttemptDetailsInfo create(
...
+                location == null ? "(unassigned)" : location.getLocationString();
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:207]: The `SubtaskExecutionAttemptDetailsInfo#host` should rename, right?

Also, the `SubtaskExecutionAttemptDetailsInfo#FIELD_NAME_HOST` and the front-end should be renamed  as well. However, I'm not sure whether the interface is called by flink users. If yes, the change is incompatible.Should we adding an extra field to store the locationString? 

Anyway, using the `host` to store `locationString` for rest api does't make sense.

**Review by 1996fanrui — CHANGES_REQUESTED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java (hunk 4 above)
Comment (X-czh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java:283]: Sure

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java (hunk 5 above)
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:207]: Other rest api has this problem as well, such as:

- JobVertexTaskManagersInfo.TaskManagersInfo#host
- SubtasksTimesInfo.SubtaskTimeInfo#host
- SubtasksAllAccumulatorsInfo.SubtaskAccumulatorsInfo#host

Comment (X-czh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:207]: You made a good point. We should by no means rename the field directly as it breaks the backward-compatibility. I propose:
1. Add a new field called "location" (already used in `JobExceptionsInfoWithHistory`) for storing the host + port info.
2. Update the front-end to adopt the new location field.
3. Rename the column name from "Host" to "Location" on the Web UI (previously we reached consensus on using "Host:port", but after a second thought, I think "Location" is more concise)
4. Keep the old "host" field untouched. Notice that the info stored in the old "host" field is inconsistent, sometimes only host is stored, sometimes host + port is stored, I'll leave them as they were with a few comments to keep being compatible.
What do you think @1996fanrui @huwh ?

Comment (huwh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:207]: Good catch. It's better to align the name between the UI and the Rest API.

We need a FLIP and discussion in dev ML before this change.

Comment (X-czh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/SubtaskExecutionAttemptDetailsInfo.java:207]: As is reminded by others, changing REST API fields would normally require a FLIP. I'll send an email on the discuss thread first and prepare one if needed.

**Comment by huwh** [comment]
closed due to duplicating with "https://issues.apache.org/jira/browse/FLINK-33146".

**Pull request closed without merging**